### DISCUSSION TITLE
The name "....FilterInput" was already registered by another type.

### DISCUSSION THREAD

**Discussion by bo55vxr** [ChilliCream/hotchocolate#5216] [Q&A]
My Graph QL sits over the top of multiple APIs and some of these APIs have classess with the same name. Consider the following..

```
namespace MyApp.CustomerServices
{
    public class Customer
    {
        [UseFiltering]
        public IEnumerable<Address> Addresses {get; set;}
    }
    public class Address
    {
        public string Addr1 {get; set;}
        public string Addr2 {get; set;}
        public string Addr3 {get; set;}
    }
}
```

```
namespace MyApp.SupplierServices
{
    public class Supplier
    {
        [UseFiltering]
        public IEnumerable<Address> Addresses {get; set;}
    }

    public class Address
    {
        public string AddrLine1 {get; set;}
        public string AddrLine2 {get; set;}
        public string AddrLine3 {get; set;}
    }
}
```

As you can see, I have two separate name spaces each with an `Address` class but slightly different. 

The start up of the Graph QL system performs with no error, however when I send in a request I then get this error.

```
The name \u0060AddressFilterInput\u0060 was already registered by another type. (HotChocolate.Data.Filters.FilterInputType\u003CMyApp.SupplierServices.Address\u003E)
```

This is obviously because there are TWO addresses classes, albeit in different namespaces, but HC does not seem to recognise that.

Other than changing the name of one of the classes, is there any way to tell HC to name the filter differently?

**Comment by sigmarsson** [comment]
Yeah and unfortunately the annotation ```[GraphQLName("")]``` does not impact either the schema generation. I have not understood yet ```Stitching```, perhaps it can be harnessed to resolve this or that would resolve this only if the different ```Address``` classes would be located in different API -s ?

**Reply by PascalSenn** [reply]
Interesting. Can you share some code? this should work

**Reply by sigmarsson** [reply]
Do you expect it to be working even in version ```11.0.9``` ? Because it does not. I apply the annotation on both classes, which are having the same name in different name spaces but to no avail. 

![image](https://user-images.githubusercontent.com/43340959/178694552-d6264cb9-000b-4b84-a6ae-3ce3a354f886.png)

![image](https://user-images.githubusercontent.com/43340959/178694843-586c3329-6301-4eae-8e2b-4cf43c205f52.png)

![image](https://user-images.githubusercontent.com/43340959/178695017-e402506b-07eb-49c1-9c08-4fb6d53dbea5.png)

**Reply by PascalSenn** [reply]
@sigmarsson  So "." is not valid in a GraphQL type name.

**Reply by sigmarsson** [reply]
But it is being replaced with an underscore in SDL? Hence the dot will not remain in place?

**Reply by sigmarsson** [reply]
@PascalSenn I have overridden the ```DefaultNamingConventions``` ```GetType(Type type)``` method and indeed, I witness the name is actually equals to the value, which is specified in ```[GraphQLName("")]``` . But this does not seem to affect the ```ListFilterInputType*``` because the error keep persisting. Why ?

**Comment by PascalSenn** [comment]
@bo55vxr  You can either override the NamingConvetions (create a subclass of DefaultNamingConvntions and override GetTypeName and register it on the schema) or you can create customer filter types and override the names. e.g. 
```
public class AddressFilterInputType: FilterInpuType<Address> 
{
   protected override void Configure(IFilterInputTypeDesdcriptor descriptor) => descriptor.Name("SomeName");
}
```
[GraphQLName] should actully also work. (apparently @sigmarsson  has issues with it though)

**Reply by sigmarsson** [reply]
The [v 11 documentation](https://chillicream.com/docs/hotchocolate/v11/fetching-data/filtering/#binding-of-filtertypes) is wrong relating to ```protected override void Configure()``` because is actually ```sealed``` .

**Reply by PascalSenn** [reply]
@sigmarsson That would be bad as this is the configuration interface :) have a look here [Source](https://github.com/ChilliCream/hotchocolate/blob/main/src/HotChocolate/Data/src/Data/Filters/FilterInputType%601.cs#L37-L39)

There are two Configure method. The generic type only seales the Confiure of the non generic one

**Reply by bo55vxr** [reply]
Bit of a dumb question, but how to I register the new `public class AddressFilterInputType: FilterInpuType<Address>` in Progam.cs?

**Reply by bo55vxr** [reply]
FYI - I am using v12

**Comment by sigmarsson** [comment]
@bo55vxr  what helped me to resolve collision among ```*FilterInputType*``` was overriding ```GetTypeName()```


`builder.Services.AddGraphQLServer().AddConvention<IFilterConvention, MyFilterConvention>();`



```
    public class MyFilterConvention : FilterConvention
    {
        public override NameString GetTypeName(Type type)
        {
            ...     
            return base.GetTypeName(type);
        }

        protected override void Configure(IFilterConventionDescriptor descriptor)
        {
            descriptor.AddDefaults();
        }
    }
```

**Comment by bo55vxr** [comment]
Hi @sigmarsson (Nice Avatar BTW)

So, using the details provided, I have created the following class

```
public class ContactDetailsConvention : FilterConvention
{
    public override NameString GetTypeName(Type type)
    {
        return base.GetTypeName(type);
    }

    protected override void Configure(IFilterConventionDescriptor descriptor)
    {
        descriptor.AddDefaults();
    }
}
```

And then registered it 

```
IRequestExecutorBuilder hc = services.AddGraphQLServer()
                                                             .AddAuthorization()
                                                             .AddFiltering()
                                                             .AddSorting()
                                                             .AddErrorFilter<ErrorFilter>()
                                                             /* These prevent Filter "collisions" where class names are the same */
                                                             .AddConvention<IFilterConvention, ContactDetailsConvention>();
```

When I run the code and make a schema request (from Postman), if then throws the following exception

 _There are two conventions registered for HotChocolate.Data.Filters.IFilterConvention in scope default. Only one convention is allowed. Use convention extensions if additional configuration is needed. Colliding conventions are HotChocolate.Data.Filters.FilterConvention and VAA.EAPI.GraphQL.Domain.CreateCustomer.Models.ContactDetailsConvention
 (VAA.EAPI.GraphQL.Domain.Loyalty.Queries.CustomerType)_

Using v12.15.0 of HC

**Comment by bo55vxr** [comment]
Well, using all the information here, and a bit of creative coding, I managed to now resolve everything I need.

1. With a custom `IFilterConvention`, I am required to remove the `.AddFiltering()` registration
2. In the custom filter, you have to manipulate the response to provide a unique name so the way I implemented this was to use a custom class attribute and decorate each of the duplicate classes. In the filter, check the _Type_ that has been passed to see if this attribute is present and if so use the name provided, else use the default name.

**Custom Filter and Attribute**
```
using HotChocolate;
using HotChocolate.Data;
using HotChocolate.Data.Filters;
using System;
using System.Linq;

namespace MyProject.Middleware
{
    [AttributeUsage(AttributeTargets.Class, AllowMultiple = false)]
    public class MyCustomFilterConventionAttribute: Attribute
    {
        public string FilterName { get; }

        public MyCustomFilterConventionAttribute(string filterName)
        {
            FilterName = filterName;
        }
    }
    public class MyCustomFilterConvention: FilterConvention
    {
        public override NameString GetTypeName(Type type)
        {
            NameString result = base.GetTypeName(type);

            var attr = type.GetCustomAttributes(false).FirstOrDefault(attr => attr is MyCustomFilterConventionAttribute);
            if (attr is not null)
            {
                result = ((MyCustomFilterConventionAttribute)attr).FilterName;
            }
            return result;
        }

        protected override void Configure(IFilterConventionDescriptor descriptor)
        {
            descriptor.AddDefaults();
        }
    }
}
```
**Registering**

```
IRequestExecutorBuilder hc = services.AddGraphQLServer()
                                     .AddAuthorization()
                                     .AddFiltering<MyCustomFilterConvention>()
                                     .AddSorting()
                                     .AddErrorFilter<ErrorFilter>();
```

**Service A Models**
```
namespace MyProject.Models.ServiceA
{
    [MyCustomFilterConventionAttribute(filterName: "ServiceAContactDetailsFilter")]
    [GraphQLName("ServiceAContactDetails")]
    public class ContactDetails
    {
        ....
    }
}
```
**Servce B Models**
```
namespace MyProject.Models.ServiceB
{
    [MyCustomFilterConventionAttribute(filterName: "ServiceBContactDetailsFilter")]
    [GraphQLName("ServiceBContactDetails")]
    public class ContactDetails
    {
        ....
    }
}
```

**Reply by bo55vxr** [reply]
@PascalSenn / @sigmarsson - Would welcome a critique review ;)

**Comment by rtomita-plana** [comment]
I had a very similar issue, but in my case, just overriding the filters wasn't enough. I had to extend the `HotChocolate.Types.Descriptors.DefaultNamingConventions` class to actually fix the issue my good:
```csharp
public class MyCustomNamingConventions : DefaultNamingConventions
{
    public override string GetTypeName(Type type)
    {
        // my custom code here.
    }

    public override string GetTypeName(Type type, TypeKind kind)
    {
        // my custom code here.
    }
}
```
The code in the class is almost identical to the original base class, except that it's intented to substitute the original `GetGraphQLName()` extension method call with a private method that again, replicates the original almost identically, just to replace a `return typeName ?? type.Name` line with some custom private method resolver that will consider part of the namespace of the class in the name: `return typeName ?? MyCustomNameResolver(type);`

And to make it work, just add `.AddConvention<INamingConventions, MyCustomNamingConventions>()` to the GraphQL extensions list and it works for all filters, sortings, ect.

**Reply by dystopiandev** [reply]
Basically:
```cs
using HotChocolate.Types.Descriptors;

public class NamingConventions : DefaultNamingConventions
{
    public override string GetTypeName(Type type, TypeKind kind)
    {
        // Read any explicit [GraphQLName] attribute first
        GraphQLNameAttribute? graphQlNameAttr = type
            .GetCustomAttributes(typeof(GraphQLNameAttribute), inherit: false)
            .OfType<GraphQLNameAttribute>()
            .FirstOrDefault();

        // If an explicit name is provided, honor it
        string? typeName = graphQlNameAttr?.Name;

        // If no explicit name, build a unique name using the last namespace segment + class name
        if (string.IsNullOrEmpty(typeName))
        {
            typeName = NamingConventions.resolveUniqueName(type);
        }

        // Let the base class apply the TypeKind suffix (e.g. "FilterInput", "SortInput")
        // by temporarily pretending we have a renamed type.
        // We do this by calling the base with a synthetic name already set.
        return base.GetTypeName(type, kind) switch
        {
            // Replace whatever name the base produced with our unique one + its suffix
            var baseName when baseName.EndsWith("FilterInput", StringComparison.InvariantCulture) => typeName +
                "FilterInput",
            var baseName when baseName.EndsWith("SortInput", StringComparison.InvariantCulture) => typeName +
                "SortInput",
            _ => typeName,
        };
    }

    // Builds "LastNamespaceSegment_ClassName" when the class has no explicit GraphQL name.
    // e.g. MyApp.CustomerServices.Address  →  CustomerServicesAddress
    //      MyApp.SupplierServices.Address  →  SupplierServicesAddress
    private static string resolveUniqueName(Type type)
    {
        string? namespacePart = type.Namespace?.Split('.').LastOrDefault(); // take the innermost namespace segment

        return string.IsNullOrEmpty(namespacePart) ? type.Name : $"{namespacePart}{type.Name}";
    }
}
```
